vortex-expr/src/pack.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::Any;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use itertools::Itertools as _;
+use vortex_array::aliases::hash_set::HashSet;
+use vortex_array::array::StructArray;
+use vortex_array::validity::Validity;
+use vortex_array::{ArrayData, IntoArrayData};
+use vortex_dtype::field::Field;
+use vortex_dtype::FieldNames;
+use vortex_error::{vortex_bail, VortexResult};
+
+use crate::{ExprRef, VortexExpr};
+
+#[derive(Debug, Clone)]
+pub struct Pack {
+    names: FieldNames,
+    values: Vec<ExprRef>,
+    validity: Option<ExprRef>,
```

Comment:
I'm not sure about this, it feels like a bit of a hammer. If you image `pack` as a SQL expression, I'd expect it to _always_ return NonNullable structs.

Unless there's a use-case for this?

Reply:
I think this is a separate operation all together, `apply_null` or something

Reply:
I imagined that we would use `Pack` to reconstruct a tree of structs from a flattened file. In particular, if we had `a: {b: u8, c:u8}, d: u8` we could store that using a columnar layout of four columns: `a.b, a.c, a.is_valid, a.d` and reconstruct it using a `Pack` "projection": `Pack("a": Pack("b", Column("a.b")), ...)`.

Joe, your approach is reasonable. I discussed this a bit [in slack](https://spiraldb.slack.com/archives/C07BV3GKAJ2/p1735937201041239). I called that operation `mask`. PyArrow calls it `replace_with_mask` whereas Pandas calls it `mask`.

In addition to `mask`, we will want: `SparseArray` or similar for applying the mask to arrays that lack ValidityMetadata and also `GetFlattenedField` or similar for accessing a field with parent's validity applied.

Reply:
Validity is gone. We can discuss the need for `mask` and friends in a different forum.